Assert the type of `global $product` before calling its methods.

tomalec opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Assert the type of `global $product` before calling its methods.
Prevent fatal error from https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/266
Add a few tests

### Checks:
<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Run automated tests
2. Test with Elementor Pro/Wait for reproduction steps for https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/266 and redo them
   > I get a fatal error when I open WP – Templates and click view. Otherwise, this error does not occur. Disabling your plugin also clears this error.



### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

> Fix - Fatal error when running with Elementor.

tomalec committed:
Assert the type of `global $product`,

before calling its methods.

Prevent fatal error from https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/266

martynmjones commented on tests/unit-tests/ProductDetail.php:
```diff
@@ -42,4 +43,92 @@ public function test_view_item_event() {
+
+		// Mimic WC action
+		do_action( 'wp_enqueue_scripts' );
+		do_action( 'woocommerce_before_single_product' );
```
`do_action( 'woocommerce_before_single_product' );` triggers [woocommerce_ouput_all_notices](https://github.com/woocommerce/woocommerce/blob/3679f019bbe4a1586e28fd9e9d2710e89786f2b1/plugins/woocommerce/includes/wc-template-functions.php#L3840) so although the tests are passing I'm getting some unexpected output when they run:

![Screenshot 2023-05-03 at 06 13 37](https://user-images.githubusercontent.com/40762232/235836942-297def18-ae08-4e02-a10b-1cfa7e43a8e1.png)

martynmjones reviewed (approved):
Hey @tomalec,

Thanks for the quick fix on this compatibility error.

I've tested this PR and it works as expected ✅ , however, I wasn't able to test it against the reported issue with Elementor Pro because no matter what I did I wasn't able to reproduce the critical error 🤔 .

As the `setup_frontend_scripts` is hooked into `woocommerce_before_single_product` it's pretty confusing as to why the global $product variable wouldn't be an instance of a WC Product at that point. This PR will surely fix the fatal error but I'm wondering if there won't be other unexpected behavior on the merchant's site and it might be an issue in Elementor instead.

That being said I'm sure this will resolve the critical so I think it's safe to merge and release. I've left one comment on the tests as I'm getting some unexpected output when I run them.

tomalec committed:
Prevent `woocommerce_before_single_product` to pollute test runner output

Addresses https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/267#discussion_r1183244855

tomalec committed:
Change order of assertions, to read `registered` prop once we asserted it.

tomalec commented:
@martynmjones could you check once again on your local unit tests?

I'm confused with unit tests results.
They pass on my local. They passed here for `97dd102`. But they [fail for `e809d49` here](https://github.com/woocommerce/woocommerce-google-analytics-integration/actions/runs/4875300752/jobs/8697356489), even though I [didn't change anything in `e809d49`](https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/267/commits/e809d49f8b51e6ba04cd1014a75263680c7410f3) regarding the script registration and anything that happens up to the line 62.

martynmjones reviewed (commented):
Hey @tomalec,

> I'm confused with unit tests results.
They pass on my local. They passed here for 97dd102. But they [fail for e809d49 here](https://github.com/woocommerce/woocommerce-google-analytics-integration/actions/runs/4875300752/jobs/8697356489), even though I [didn't change anything in e809d49](https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/267/commits/e809d49f8b51e6ba04cd1014a75263680c7410f3) regarding the script registration and anything that happens up to the line 62.

Yeah that's pretty confusing 🤔 the changes look good and tests are all passing for me locally as well.

Maybe reverting the changes and removing the `woocommerce_ouput_all_notices` hook from `woocommerce_before_single_product` instead of using `ob_get_clean()` would resolve the ouput on the test runner and have them pass on Github?

That being said, I have absolutely no idea why using an output buffer would cause these tests to fail, especially considering the fact that one of the tests wasn't changed 😕

tomalec committed:
Merge branch 'trunk' into fix/266-product-bool-fatal

tomalec commented:
I tried reverting changes from https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/267/commits/e809d49f8b51e6ba04cd1014a75263680c7410f3 and it was still failing :confused: 
I'll merge `trunk` maybe some travis/build changes from it affects the GHA :shrug:

tomalec commented:
Still fails, I even did a new PR, with the very same revision [that passed here](https://github.com/woocommerce/woocommerce-google-analytics-integration/actions/runs/4844772766/jobs/8697914589). There it fails :confused: 
https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/272

martynmjones commented on tests/unit-tests/ProductDetail.php:
```diff
@@ -42,4 +43,98 @@ public function test_view_item_event() {
+		do_action( 'wp_enqueue_scripts' );
+
+		// Assert the handle is regeistered with the correct name, but not yet enqueued.
+		$this->assertEquals( 'woocommerce-google-analytics-integration', $gtag->script_handle, '`WC_Google_Gtag_JS->script_handle` is not equal `woocommerce-google-analytics-integration`' );
```
There were some changes in `trunk` that have been merged back into this branch so the handle is no longer `$gtag->script_handle` and is now `$gtag->script_handle . '-ga-integration'`.

I just ran this test on a [new branch](https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/273) after changing the script handle and they passed.

Really confusing though as if you look back at the previous commits on this PR those changes _definitely_ didn't exist on the branch when they first failed but it definitely seems like that is why it was failing. I have no idea what was happening there. 😕 

The timing also "makes sense" as #269 was merged into `trunk` 1 day after you opened this PR so the tests passed up until the changes were made in `trunk`.

puntope replied:
For context: 

The reason for this change is because now we are adding more than one script in `WC_Google_Gtag_JS`

`$this->script_handle . '-ga-integration'` (former  `$this->script_handle`) and `$this->script_handle . '-actions'`

tomalec committed:
Move `WC_Google_Gtag_JS` tests to a separate class

as we will add more and more to check scripts registration behavior.

tomalec committed:
Add tests for enqueueing `actions.js`

Adapt `gh-integration.js` tests for new paths and names.
Addresses https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/267#discussion_r1185731468

tomalec commented:
> There were some changes in `trunk` that have been merged back into this branch so the handle is no longer `$gtag->script_handle` and is now `$gtag->script_handle . '-ga-integration'`. …

Thanks. It's still a mystery to me why that change in `trunk` affected this branch before merging :confused: 

----

I added `-actions` to the test, also moved them to a separate file, and rephrased a little. Would you mind another look? If not, I'd merge it today EOD, to make sure no other mystery will break them ;) We could do potential improvements in a separate tests-only PR, but ship that fix for merchants next week
.

martynmjones reviewed (approved):
> I added -actions to the test, also moved them to a separate file, and rephrased a little. Would you mind another look? If not, I'd merge it today EOD, to make sure no other mystery will break them ;) We could do potential improvements in a separate tests-only PR, but ship that fix for merchants next week

Mysteries are not for Fridays. 😅 

I tested the PR again with the latest changes and the tests are passing locally (_still_) so LGTM ✅
